### ISSUE TITLE
Handshake error in client (Os error CERTIFICATE_VERIFY_FAILED: certificate has expired (handshake.cc:393))

### ISSUE THREAD

**Issue by fantoche7** [thunder-app/thunder#178]
**Description**  
unable to connect , I beleive It has something to do with flutter on older android devices

**Expected Behavior**  
be able to connect 

**Device & App Version:**  
 - OS: android
 - Version: 5.1

**Comment by hjiangsu** [comment]
Could you provide me with some more details about this issue? That'll help me figure out the root cause better!

For example:
- OS version
- Instance you're trying to connect to
- Are you logged in, or does it simply show when you start the app

**Comment by fantoche7** [comment]
OS : android 5.1
The error show at startup ,and when ere trying to do something, like searching, I bet the app can't make any connetion to the internet.

**Comment by ddCeka** [comment]
I have the same problem.
OS: Android 10

The problem(handshake not resolved) shows when I tried to use the app routing it with adguard. While I excluded it from adguard, it works as usual. Look like adguard certificate causing this?

**Comment by hjiangsu** [comment]
Hmm, interesting - I'm not too sure what it could be. Do you have Sentry enabled in the settings (this is only because you mentioned adguard which might be blocking Sentry)? Try to disable Sentry if its enabled and see if that changes anything

The certificate errors usually seem to indicate that there was an issue with checking for the instance's credentials (through HTTPS)

Could you let me know the instances that you've tried to connect to? I can try to see if I can replicate it on my end

**Comment by ddCeka** [comment]
> Oh hmm, I have to go through an application process in order to sign up with your instance. I have tried to set the default instance temporarily to `lemmy.dbzer0.com` and have not yet encountered an issue (at least on the iPhone simulator)
> 
> ![simulator_screenshot_4293345F-1069-4B10-B4D4-110D10C1B148](https://user-images.githubusercontent.com/30667958/250582448-beaeb94b-f334-473f-849d-5bf302332af7.png)

Yeah, at the moment the only way I do just disabling HTTPS filtering on adguard for this and work just fine

**Comment by hjiangsu** [comment]
I'm looking into this: https://adguard.com/kb/general/https-filtering/what-is-https-filtering/#https-filtering-issues-on-android-7

It might be possible for Thunder to trust user-installed certificates. I would have to take a deeper look into it but this might be the root cause of the issue

**Comment by hjiangsu** [comment]
So I changed some configs to allow user-installed certificates, but I'm not sure if it works. If you don't mind, could you test out this build and let me know if it works? (It's uploaded on google drive because GitHub doesn't allow me to upload APKs or files above 25MB)

https://drive.google.com/file/d/17BJBobYkD3U7_UBktNmdorngiQvJVM6E/view?usp=sharing

**Comment by MetehanOzyurek** [comment]
For devices under Android 8, check this: https://github.com/dart-lang/http/issues/627

and also: https://lemmy.world/post/989618

**Comment by hjiangsu** [comment]
Hmm, I'll take a look at that later! Thanks for the links, I appreciate it

**Comment by ddCeka** [comment]
> So I changed some configs to allow user-installed certificates, but I'm not sure if it works. If you don't mind, could you test out this build and let me know if it works? (It's uploaded on google drive because GitHub doesn't allow me to upload APKs or files above 25MB)
> 
> https://drive.google.com/file/d/17BJBobYkD3U7_UBktNmdorngiQvJVM6E/view?usp=sharing
> 
> Edit: This is the VirusTotal scan also, in case you wanted to be cautious (you can also perform it yourself if you wish!) https://www.virustotal.com/gui/file/828fe058b4591968122ce96f06a70ebb9cebfccd7e75e9d34ceb31b1a45ebade/detection

Hi, I tried the app from shared drive and it still persists with certificate verify failed.

**Comment by ajsosa** [comment]
@hjiangsu Hmm I'm not sure how I feel about it but it looks like we should be able to modify lemmy-api-client with this work around.

https://stackoverflow.com/questions/69511057/flutter-on-android-7-certificate-verify-failed-with-letsencrypt-ssl-cert-after-s

Would have to verify the cert is actually legit.

**Comment by bazsy01** [comment]
Just tried out the solution from https://stackoverflow.com/a/69481863 [here](https://github.com/bazsy01/thunder/commit/6d443e5c7f25a6ce6d39f58641ce4b076f284800).
This resolved my issue with lemmy.world since that uses letsencrypt cert but it is still a painfully manual workaround. There might be instances relying on other root CAs, and including them all would also increase maintenance as they expire.

Native Android apps usually rely on [GMS to update SSL](https://developer.android.com/training/articles/security-gms-provider) compatibility. Is there a way in Flutter to provide a more general solution?

**Comment by hjiangsu** [comment]
> Just tried out the solution from https://stackoverflow.com/a/69481863 here.

Hmm, I wonder if we can use this solution as a stop gap for now, and revisit this in the future

If we wanted to, I think we could modify the Android code to potentially support the solution you linked. But keep in mind that not everyone uses Google Play Services so this solution would exclude those users.

**Comment by daffydock** [comment]
I have the same issue on 7.0
Google Play Services and all Google related APKs have been removed.

**Comment by Om-Leocan** [comment]
Any update ? 
Are you able to solve this ? 
Because i am getting the same with android 7 .

**Comment by hjiangsu** [comment]
@Om-Leocan I've created a PR which should solve this issue in #1675, and should land in the general release for v0.6.0!
